synapse/handlers/presence.py

```diff
@@ -224,7 +225,9 @@ async def current_state_for_users(
                 states.update(new)
                 self.user_to_current_state.update(new)
 
-        return states
+        # mypy does not realize that is Dict[str, UserPresenceState]]
+        # and not Dict[str, Optional[UserPresenceState]]]
+        return cast(Dict[str, UserPresenceState], states)
```

Comment:
I'm not surprised, given the code as it stands is convoluted. TBH I'd be tempted to rewrite it as something like:

```python
states = {}
for user_id in user_ids:
    state = self.user_to_current_state.get(user_id, None)
    if state:
        states[user_id] = state
    else:
        ...
```

Reply:
Hopefully it's better now. The database query for a single user is not implemented `_get_presence_for_user`. My thoughts were that calling `get_presence_for_users` for every single user becomes a lot of overhead.
The loops are getting smaller now and perhaps increase the performance.